test/controllers/AccountControllerSpec.scala

```diff
@@ -0,0 +1,125 @@
+      implicit val encryptedStringFormats = JsonStringEncryption
+      implicit val decryptedStringFormats = JsonStringDecryption
+      implicit val format = Json.format[LoginDetails]
+
+      val csrfToken = "csrfToken" -> SignedTokenProvider.generateToken
+      val authToken = SessionKeys.authToken -> "some-bearer-token"
+      val userToken = GatekeeperSessionKeys.LoggedInUser -> "userName"
+
+      val aLoggedInRequest = FakeRequest().withSession(csrfToken, authToken, userToken)
+      val aLoggedOutRequest = FakeRequest().withSession(csrfToken)
+
+    }
+
+    "loginPage" should {
+      "be loaded with not authenticated user" in new Setup {
+        val result = await(underTest.loginPage()(aLoggedOutRequest))
+        status(result) shouldBe 200
+      }
+
+      "be skipped with an authenticated user" in new Setup {
+        val result = await(underTest.loginPage()(aLoggedInRequest))
+        status(result) shouldBe 303
+      }
+    }
+
+    "authenticateAction" should {
+      "go to dashboard and set cookie when user authenticated successfully" in new Setup {
+        val loginDetails = LoginDetails("userName", Protected("password"))
+        val aValidFormJson = Json.toJson(loginDetails)
+        val successfulAuthentication = SuccessfulAuthentication(BearerToken("bearer-token", DateTime.now().plusMinutes(10)), "userName", None)
+
+        given(underTest.authConnector.login(any[LoginDetails])(any[HeaderCarrier])).willReturn(Future.successful(successfulAuthentication))
+
+        val result = await(underTest.authenticateAction()(
+          aLoggedOutRequest.withJsonBody(aValidFormJson)))
+        status(result) shouldBe 303
+        result.header.headers should contain("Location" -> "/api-gatekeeper/dashboard")
+        result.header.headers.get("Set-Cookie") shouldBe defined
+
+        session(result).get(SessionKeys.authToken) shouldBe Some("bearer-token")
+
+        val cookie = result.header.headers.get("Set-Cookie").get
+        cookie should include("mdtp=")
+        cookie should include("authToken=bearer-token")
+        cookie should include("LoggedInUser=userName")
+      }
+
+      "give 400 when an invalid login form is posted" in new Setup {
+        val result = await(underTest.authenticateAction()(
+          aLoggedOutRequest.withJsonBody(Json.toJson(new LoginDetails("", Protected("password"))))))
+        status(result) shouldBe 400
+      }
+
+      "go to login page if user failed to authenticate" in new Setup {
+        val loginDetails = LoginDetails("userName", Protected("password"))
+        val aValidFormJson = Json.toJson(loginDetails)
+        given(underTest.authConnector.login(any[LoginDetails])(any[HeaderCarrier])).willReturn(Future.failed(new InvalidCredentials))
+
+        val result = await(underTest.authenticateAction()(
+          aLoggedOutRequest.withJsonBody(aValidFormJson)))
+
+        status(result) shouldBe 401
+        bodyOf(result) should include("Invalid user ID or password. Try again.")
```

Comment:
If we're not using selenium for this, would it be worth using jsoup to check that the text is in an error field ?